apps/yggdrasil/lib/yggdrasil/user.ex

```diff
@@ -29,14 +36,133 @@ defmodule Yggdrasil.User do
+    role = Repo.get_by!(Role, name: role_name)
+
+    user_role = UserRole.changeset(%UserRole{},
+                                   %{user_id: user.id, role_id: role.id})
+    case Repo.insert(user_role) do
+      {:ok, _role} -> :ok
+      error        -> error
+    end
+  end
+
+  def load_permissions(user) do
+    query = from ur in UserRole,
+            where: ur.user_id ==  ^user.id,
+            preload: [role: [:role_permissions]]
+
+    permissions = query
+    |> Repo.all()
+    |> Enum.flat_map(fn ur -> ur.role.role_permissions end)
+    |> Enum.group_by(fn rp -> rp.resource end)
+    |> Enum.map(fn {r, rps} -> {String.to_atom(r), Enum.map(rps, fn rp -> String.to_atom(rp.permission) end)} end)
+    |> Enum.into(%{})
+
+    %{ user | permissions: permissions}
+  end
+
+  @doc ~S"""
+  Checks the list of resource/permissions pair given against what the user
+  has been granted and returns true if and only if all permissions supplied
+  have been granted for the user.
+
+  ## roles
+  A user can have one or more roles with each role having a set of resources
+  and a set of resources and those resources having a set of permissions. Roles
+  are not checked directly and only serve as a grouping mechanism for resources.
+
+  Since the user can have multiple roles, the permission set checked for a given
+  resource is the unique set across all the roles for that resource.
+
+  So if the user has `Role1` and `Role2` defined as follows:
+
+  * Role1
+    * foo
+      * :read
+      * :write
+  * Role2
+    * foo
+      * :read
+      * :all
+
+  The set of permissions checked for foo will be `[:read, :write, :all]`
+
+  ## Examples
+      # checking a single resource
+      User.is_granted? user, foo: [:read]
+
+      # checking multiple resources is allowed as well
+      User.is_granted? user, foo: [:read], bar: [:read, :write]
   """
-  def hash_password(changeset = %{:valid? => false}) do
+  def is_granted?(_user, []) do
+    {:error, "res_perm list was empty"}
+  end
+
+  def is_granted?(user, res_perms) do
+    results = Enum.map res_perms, fn {res, perms} ->
+      res_perm_set = user.permissions[res]
+
+      # nil is returned if key doesn't exist
+      # nil can't be turned into a map so this
+      # must be explicitly handled.
+      if res_perm_set do
+        res_perm_set = MapSet.new(res_perm_set)
+
+        # turn perm atoms into strings
+        perm_set = MapSet.new(perms)
+
+        cond do
+          MapSet.size(perm_set) == 0 ->
+            {:error, "permission set for resource #{res} can't be empty"}
+          MapSet.size(res_perm_set) == 0 ->
+            {:error, "supplied permission set to check for resource #{res} can't be empty"}
+          true ->
+            MapSet.subset?(perm_set, res_perm_set)
+        end
+      else
+        false
+      end
+    end
+
+    errors = results
+    |> Enum.filter(fn {:error, _} -> true
+                      _           -> false end)
+    |> Enum.map(fn {_, msg} -> msg end)
```

Comment:
@copenhas I'm not sure of a better way to do this. Originally I was doing Enum.any? looking for errors in the case below, but bit awkward and if one is found then I have to iterate over it again to get the errors so was like ok just build the list if it's has some elements then error out.